File: curso_c_cpp/cpp/06_herencia_polimorfismo/ejemplos/animales.cpp
/*
 * Herencia + polimorfismo con funciones virtuales.
 */
#include <iostream>
#include <string>
using namespace std;

class Animal {
protected:
    string nombre;
public:
    Animal(const string &n) : nombre(n) {}
    virtual void hacerSonido() {              // virtual: polimorfico
        cout << nombre << ": sonido generico\n";
    }
    virtual ~Animal() {}                      // destructor virtual
};

class Perro : public Animal {
public:
    Perro(const string &n) : Animal(n) {}
    void hacerSonido() override { cout << nombre << ": Guau\n"; }
};

class Gato : public Animal {
public:
    Gato(const string &n) : Animal(n) {}
    void hacerSonido() override { cout << nombre << ": Miau\n"; }
};

int main() {
    Animal *animales[3];
    animales[0] = new Perro("Rex");
    animales[1] = new Gato("Michi");
    animales[2] = new Animal("Bestia");

    for (int i = 0; i < 3; i++) {
        animales[i]->hacerSonido();          // cada uno responde segun su tipo real
    }
    for (int i = 0; i < 3; i++) delete animales[i];   // liberar
    return 0;
}

File: curso_java/12_herencia_polimorfismo/ejemplos/Animales.java
/*
 * Herencia (extends, super) y polimorfismo con @Override.
 * La clase publica (Animales) tiene el main; las demas van en el mismo archivo.
 */
class Animal {
    protected String nombre;

    public Animal(String nombre) {
        this.nombre = nombre;
    }
    public void hacerSonido() {
        System.out.println(nombre + ": sonido generico");
    }
}

class Perro extends Animal {
    public Perro(String nombre) { super(nombre); }
    @Override
    public void hacerSonido() { System.out.println(nombre + ": Guau"); }
}

class Gato extends Animal {
    public Gato(String nombre) { super(nombre); }
    @Override
    public void hacerSonido() { System.out.println(nombre + ": Miau"); }
}

public class Animales {
    public static void main(String[] args) {
        Animal[] animales = {
            new Perro("Rex"),
            new Gato("Michi"),
            new Animal("Bestia")
        };
        for (Animal a : animales) {
            a.hacerSonido();     // cada uno responde segun su tipo real
        }
    }
}
